crmsh/join_lock.py

```diff
@@ -0,0 +1,167 @@
+
+    def _run(self, cmd):
+        """
+        Run command on target node, consider specific exceptions
+        """
+        cmd_with_ssh = "ssh {} root@{} \"{}\"".format(self.SSH_OPTION, self.target_node, cmd)
+        rc, out, err = utils.get_stdout_stderr(cmd_with_ssh)
+        if rc != 0 and err:
+            if re.search("tail: cannot open .*: No such file or directory", err):
+                raise FileNotFoundError(err)
+            elif re.search("mkdir: cannot create directory .*: File exists", err):
+                raise FileExistsError(err)
+            elif re.search("ssh: connect to .*: Connection timed out", err):
+                raise TimeoutError(err)
+            else:
+                raise RuntimeError(err)
+        return rc, out
+
+    def _get_last_timestamp(self):
+        """
+        Tail the last line in lock file
+        """
+        result = {}
+        rc, out = self._run(self.TAIL_CMD)
+        if rc == 0 and out:
+            try:
+                locked_node, timestamp = out.split()
+            except ValueError:
+                raise ValueError("Invalid format for {}(should be \"node timestamp\")".format(self.JOIN_LOCK_FILE))
+            else:
+                result['node'] = locked_node
+                result['timestamp'] = timestamp
+        return result
+
+    def _remove_lock_dir(self):
+        """
+        Remove lock directory
+        """
+        self._run(self.RM_CMD)
+
+    def _create_lock_dir(self):
+        """
+        Create lock directory, mkdir command was atomic
+        """
+        try:
+            self._run(self.MKDIR_CMD)
+        except FileExistsError:
+            return False
+        return True
+
+    def _lock_or_wait(self):
+        """
+        Wait 15 seconds while detect target node has lock directory
+        """
+        file_not_found_once = False
+        warned_once = False
+        origin_timestamp = None
+
+        while not self._create_lock_dir():
+
+            try:
+                result = self._get_last_timestamp()
+            # consider lock file might not be created yet
+            except FileNotFoundError:
+                if file_not_found_once:
+                    self._remove_lock_dir()
+                    break
+                else:
+                    file_not_found_once = True
+                    time.sleep(self.DETECT_INTERVAL)
+                    continue
+
+            locked_node, timestamp = result.values()
+            if origin_timestamp:
+                res = int(timestamp) - int(origin_timestamp)
+                if res < self.UPDATE_INTERVAL:
+                    self._remove_lock_dir()
+                    break
+            origin_timestamp = timestamp
+
+            if not warned_once:
+                warned_once = True
+                bootstrap.warn("Node {} still joining, wait...".format(locked_node))
+
+            time.sleep(self.DETECT_INTERVAL)
+
+    def _update_timestamp(self):
+        """
```

Comment:
as complete outsider of the project and rfc, I think it would be useful to have the rationale of the update timestamp function.

As I do understand right now, I think the update timestamp is used in a way to see if the Connection is still alive right?